cupy/indexing/generate.py

```diff
@@ -1,9 +1,160 @@
 # flake8: NOQA
 # "flake8: NOQA" to suppress warning "H104  File contains nothing but comments"
+import cupy
+
+import six
+
+
+class AxisConcatenator(object):
+    """Translates slice objects to concatenation along an axis.
+
+    For detailed documentation on usage, see `r_`.
+
+    """
+    def _retval(self, res):
+        if self.matrix:
+            oldndim = res.ndim
+            res = makemat(res)
+            if oldndim == 1 and self.col:
+                res = res.T
+        self.axis = self._axis
+        self.matrix = self._matrix
+        self.col = 0
+        return res
+
+    def _output_obj(self, newobj, tempobj, ndmin, trans1d):
+        k2 = ndmin - tempobj.ndim
+        if (trans1d < 0):
+            trans1d += k2 + 1
+        defaxes = list(six.moves.range(ndmin))
+        k1 = trans1d
+        axes = defaxes[:k1] + defaxes[k2:] + \
+               defaxes[k1:k2]
+        newobj = newobj.transpose(axes)
+        return newobj
+
+    def __init__(self, axis=0, matrix=False, ndmin=1, trans1d=-1):
+        self._axis = axis
+        self._matrix = matrix
+        self.axis = axis
+        self.matrix = matrix
+        self.col = 0
+        self.trans1d = trans1d
+        self.ndmin = ndmin
+
+    def __getitem__(self, key):
+        trans1d = self.trans1d
+        ndmin = self.ndmin
+        objs = []
+        if isinstance(key, str):
+            return NotImplemented
+        if not isinstance(key, tuple):
+            key = (key,)
+
+        for k in six.moves.range(len(key)):
+            if isinstance(key[k], slice):
+                return NotImplemented
+            elif isinstance(key[k], str):
+                if k != 0:
+                    raise ValueError(
+                    'special directives must be the first entry.')
+                return NotImplemented
+            else:
+                newobj = key[k]
+                if ndmin > 1:
+                    tempobj = cupy.array(newobj, copy=False)
+                    newobj = cupy.array(newobj, copy=False, ndmin=ndmin)
+                    if trans1d != -1 and tempobj.ndim < ndmin:
+                        newobj = self._output_obj(newobj, ndmin, trans1d)
+                    del tempobj
+                elif ndmin == 1:
+                    tempobj = cupy.array(newobj, copy=False)
+                    newobj = cupy.array(newobj, copy=False, ndmin=ndmin)
+                    if tempobj.ndim < ndmin:
+                        newobj = self._output_obj(newobj, tempobj, ndmin, trans1d)
+                    del tempobj
+
+            objs.append(newobj)
+
+        res = cupy.concatenate(tuple(objs), axis=self.axis)
+        return self._retval(res)
+
+    def __len__(self):
+        return 0
+
+class CClass(AxisConcatenator):
+    def __init__(self):
+        """Translates slice objects to concatenation along the second axis.
+
+        This is short-hand for ``np.r_['-1,2,0', index expression]``, which is
```

Comment:
To be precise, `cupy.r_` does not support the case where the argument of `__getitem__` is `str`.